FAIR-3887: Fix various errors when running scripts

RossBarnie opened the pull request:
Only API-related change was the dictionary lookup change where we now specify `type` and `options`, rather than specifying the options at the top-level of a lookup.

Commit:
FAIR-3887: Separate authorized and unauthorized headers

Commit:
FAIR-3887: Run autopep8 across all files

Commit:
FAIR-3887: Improve usage message on upload

Commit:
FAIR-3887: Fix bad import

Commit:
FAIR-3887: Fix dataset upload example

Dictionary lookups now must specify their type separately from the
available options for that field.

Commit:
FAIR-3887: Fix errors on dataset update

Commit:
FAIR-3887: Fix code not valid in dictionary patch

Commit:
FAIR-3887: Fix requests download report failure

Commit:
FAIR-3887: Various documentation updates

Commit:
FAIR-3887: Add requirements file for python libs

Easier to run `pip install -r requirements.txt` than it is to have to
track down individual package names. `tusclient` is a particular problem
since the package name is `tuspy`. This avoids confusion, for those more
familiar with python as a language.

seantc commented on README.md:
```diff
@@ -318,13 +316,13 @@ This section describes how the update cycle in FAIR works.
 
-1. Start by creating a JSON file conforming to the metadata specification set out in [Create a Dataset Entry](#). We will call this `dataset.json`. Post this to the API and make a note of the dataset_code.
-2. Make a change to the metadata in `dataset.json`, for example, add or remove a dictionary from the `dictionaries` list, or change some values in the `catalogue` attributes.
+1. Start by creating a JSON file conforming to the metadata specification set out in [Create a Dataset Entry](#create-a-dataset-entry-api). Post this to the API and make a note of the dataset_code. For an existing dataset you can use the [`fair-api-datasets-get.py` script](fair-api-datasets-get.py), though you will have to remove some of the additional fields from this json. With either method, w ...
```
Maybe separate these into 1(a) 1(b) options

seantc commented on README.md:
````diff
@@ -378,6 +376,108 @@ To delete your test dataset, use the script [`fair-api-datasets-delete.py`](fair
+}
+```
+
+### Update a Dictionary
````
Do we need a section for updating a catalogue too?

RossBarnie replied:
Catalogues are part of the dataset update for now. Given the catalogue templating feature won't be released this year I'm happy to leave out those changes until then.

seantc commented on common/auth.py:
```diff
@@ -0,0 +1,27 @@
+from common.constants import BASE_HEADERS
+
+# Ensure we have set the token before setting the constant for use elsewhere
+def set_token():
```
Isn't this get_token()?

RossBarnie replied:
I was using `set` because the function sets the value of the constant. Admittedly it's unintentionally a public method. Reading the code again I think it makes more sense to use these functions directly, rather than using the constant which has call the function anyway.

RossBarnie replied:
Scratch the above, changing the function name to get is sufficient given the other uses in the constants file.

seantc replied:
Were you updating this one?

seantc commented on common/auth.py:
```diff
@@ -0,0 +1,27 @@
+FAIR_API_TOKEN = set_token()
+
+
+def set_authorized_header():
```
get_ again I think

seantc commented on fair-api-datasets-get.py:
```diff
@@ -0,0 +1,23 @@
+    error_data = response.json()
+    print(
+        f'Failed to retrieve Dataset. Status code: {response.status_code}, Error message: {error_data["error"]["message"]}')
+    exit(2)
```
Think this should be an exit status 1

RossBarnie replied:
I made it exit status 2 because the convention in other files is to use exit status 1 for missing arguments, token, or endpoint. Exit status 2 means that a response failed. I can make this more explicit.

seantc commented on fair-api-datasets-update.py:
```diff
@@ -3,58 +3,66 @@
     diff = DiffHelper.dataset_diff(original, data)
-    print (f'\nPATCH {dataset_url(dataset_code)} --data {json.dumps(diff, indent=2)}')
+    print(
+        f'\nPATCH {dataset_url(dataset_code)} --data {json.dumps(diff, indent=2)}')
```
Is this meant to show you the curl to use? Would be nice if it was prefixed with `curl` so this is a runnable command

RossBarnie replied:
I didn't write this initially but that seems to be the intention, I'll add it in.

seantc reviewed (commented):
Couple of comments. Looks good otherwise.

Commit:
FAIR-3887: Split patch creation instructions

Commit:
FAIR-3887: Rename set_ functions to get_

Commit:
FAIR-3887: Make exit code convention explicit

Commit:
FAIR-3887: Consistently curl-ify requests

Commit:
FAIR-3887: Fix dry-run argument assignment

Commit:
FAIR-3887: Fix missed set_ to get_ function

Commit:
FAIR-3887: Fix ambiguous dataset update arguments

Unclear whether dataset code required or not. Dataset identifier had
been determined from the patch file but this doesn't match the API so
now matches other scripts and the API more closely by expecting the user
to provide the code in the script arguments.

Commit:
FAIR-3887: Remove py file executable permission

We expect users to execute these files with `python` as the executable,
rather than executing the files directly, as noted in the README.

Commit:
FAIR-3887: Fix bad usage message

Also adds a comment specifying we *must* add comments to patch
definitions.
